### PR TITLE
fix(#99): proof of concept

### PR DESCRIPTION
**Pull request by hgwood**
`JSON.stringify` does the job of quoting and escaping. What do you think? Please submit any case you might think of.

ping @ebriand

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -11,7 +11,9 @@ function crossEnv(args) {
   if (command) {
     const proc = spawn(
       commandConvert(command),
-      commandArgs.map(commandConvert),
+      commandArgs
+        .map(commandConvert)
+        .map(arg => arg.match(/[ "']/) ? JSON.stringify(arg) : arg),
```
Comment (hgwood) [inline src/index.js]: I extracted this into a function but forgot to push. Pending.

File: src/index.test.js [hunk 2]
```diff
@@ -57,6 +57,58 @@ it(`should handle quoted scripts`, () => {
...
+  expect(crossSpawnMock.spawn).toHaveBeenCalledWith(
+    'cat',
+    ['"/path\\"with\\"quotes"'],
+    {
+      stdio: 'inherit',
+      shell: true,
+      env: process.env,
+    },
+  )
+})
```
Comment (hgwood) [inline src/index.test.js]: I need to add a case where the command it self has spaces. It shouldn't be quoted.

**Comment by hgwood** [comment]
This is published as `@hgwood/cross-env@fix99`. @ebriand, care to try it?

**Comment by hgwood** [comment]
Looking back at #100, I'm having doubts. The command there is (ignoring env setters) `"cross-env nyc mocha --require test/helper.js \"test/spec/**/*.test.+(js|jsx)\""`. So cross-env gets `['nyc', 'mocha', '--require', 'test/helper.js', 'test/spec/**/*.test.+(js|jsx)']`. How can cross-env know that the user wishes that this last argument be quoted? Maybe they'd like it to be expanded. cross-env can't know, can it? Am I missing something? Or was my instinct right in the first place and #99 is not a bug?

**Comment by hgwood** [comment]
Maybe we should not use the `shell` option after all. If we think about it, the OS shell does one pass on the user's command before running cross-env. Then using the `shell` option triggers *another* pass, when the user would expect only one. That's why we are having this issue, aren't we?

**Comment by kentcdodds** [comment]
I'm kinda thinking that as well @hgwood... So how can we solve the original problem? Should we use a different package to spawn our commands?

**Comment by hgwood** [comment]
I need to investigate more but I'm guessing the reason cross-spawn can't do what we want (that is, quoted scripts) is because [it escapes "dangerous" commands](https://github.com/IndigoUnited/node-cross-spawn/blob/master/lib/util/escapeCommand.js). "dangerous" is defined as not matching `/^[a-z0-9_-]+$/i`. cross-env needs the command to not be escaped. My question now is: why does cross-spawn have this protection? Do you know?

**Comment by kentcdodds** [comment]
I don't know :-/

**Comment by hgwood** [comment]
[Possibly interesting](https://github.com/IndigoUnited/node-cross-spawn/pull/5#issuecomment-52182057). It says `exec` would be more appropriate. But `exec` has the drawback of buffering stdio (see [here](https://www.hacksparrow.com/difference-between-spawn-and-exec-of-node-js-child_process.html)).

**Comment by hgwood** [comment]
Or maybe the `shell` option should only be enabled if the command passed to cross-env is a quoted script. Following question is how to detect a quoted script? Maybe it contains spaces.

**Comment by kentcdodds** [comment]
Yeah, definitely don't want to use `exec` due to the buffering stdio thing. What about switching to `spawn-command` like I did in #77? Would that work?

**Comment by hgwood** [comment]
I have looked at the spawn-command code and as far as I can tell it does the same thing as the `shell` option.

**Comment by hgwood** [comment]
Here is how I see it right now.

As the user of cross-env, I want one of two things.

- Either my command is *not* quoted, and I want the shell to process it *one* time.
- Or my command *is* quoted, and I want the shell to process it *twice* (one time to remove the quotes and one other to actually run it).

When I say "command" here, I'm excluding arguments. For example:

- In `echo "a"`, "echo" is the command and it's not quoted.
- In `"echo a"`, "echo a" is the command and it's quoted.

If I'm correct, what cross-env should do is enable `shell` only in the second case. Now how can cross-env decide which case it is running? I'm not sure.

**Comment by hgwood** [comment]
Maybe cross-env could 2 bins? That would let the user decide and cross-env would not have to guess (and potentially fail).

**Comment by hgwood** [comment]
Having 2 bins would also avoid the breaking change. :)

**Comment by kentcdodds** [comment]
I like the idea of 2 bins. But I want to preserve the original use-case with the main bin:

```
cross-env NODE_ENV=production webpack --config build/webpack.config.js
```

So, as long as we can preserve that then I'm good.

**Comment by hgwood** [comment]
Just to be sure we're on the same page, this particular use case would work with both bins. Differences show up only when there are escape sequences or quotes.

I agree the "right" thing would be to have to main bin do the v3 behavior and to other one the v4. What I was afraid of is that some users that have or will upgrade to v4 by changing their scripts, then will discover they have to revert everything for v5, but actually they would "just" have to use the other bin. Still a bit annoying. To avoid this situation, I suggest that we deprecate v4 on npm so that people don't upgrade to it and wait for v5. What do you think?

**Comment by kentcdodds** [comment]
Closed in favor of #104

**Pull request closed without merging**